### PR TITLE
Test: Model loaders as before each and restore mocks properly

### PR DESCRIPTION
**Pull request by kasperpeulen**
## What I did

Model loaders as before each and restore mocks properly

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

* Add console.log in preview, meta and story loaders. And check that they are run in that order, even if you gonna await something before it.
* Setup a spyOn in a play function, and see that the mock is reset when you are switching to a new story.

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This pull request has been released as version [`0.0.0-pr-24948-sha-12960c76`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24948-sha-12960c76). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-24948-sha-12960c76`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24948-sha-12960c76) |
| **Triggered by** | @kasperpeulen |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`kasper/model-loaders-as-before-each`](https://github.com/storybookjs/storybook/tree/kasper/model-loaders-as-before-each) |
| **Commit** | [`12960c76`](https://github.com/storybookjs/storybook/commit/12960c76f2aaa56b01c1e9c3c13114e5c5b5542d) |
| **Datetime** | Wed Nov 22 16:24:00 UTC 2023 (`1700670240`) |
| **Workflow run** | [6960070947](https://github.com/storybookjs/storybook/actions/runs/6960070947) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=24948`_
</details>

### CODE REVIEW

**Review by shilman — COMMENTED** [review]
Great job pulling this together @kasperpeulen. What's the release plan with `@storybook/csf`?

File: code/lib/types/src/modules/story.ts [hunk 1]
```diff
@@ -42,30 +47,40 @@ export type ProjectAnnotations<TRenderer extends Renderer> = CsfProjectAnnotatio
...
 };
 
-export type NormalizedProjectAnnotations<TRenderer extends Renderer = Renderer> =
-  ProjectAnnotations<TRenderer> & {
-    argTypes?: StrictArgTypes;
-    globalTypes?: StrictGlobalTypes;
-  };
+export type NormalizedProjectAnnotations<TRenderer extends Renderer = Renderer> = Omit<
+  ProjectAnnotations<TRenderer>,
+  'decorators' | 'loaders'
```
Comment (yannbf) [inline code/lib/types/src/modules/story.ts:52]: why did you need to omit and reintroduce these types right after?

File: code/lib/test/src/index.ts [hunk 2]
```diff
@@ -32,3 +25,16 @@ export const { expect } = instrument(
...
+  if (parameters?.test?.mockReset === true) {
+    resetAllMocks();
+  } else if (parameters?.test?.clearMocks === true) {
+    clearAllMocks();
+  } else if (parameters?.test?.restoreMocks !== false) {
+    restoreAllMocks();
+  }
+};
+
+// @ts-expect-error Wwe are using this as a default loader, if the test package is used. But we don't want to get into optional peer dep shenanigans.
```
Comment (yannbf) [inline code/lib/test/src/index.ts:39]:
```suggestion
// @ts-expect-error We are using this as a default Storybook loader, when the test package is used. This avoids the need of optional peer dependency workarounds.
```

File: code/lib/test/src/index.ts [hunk 3]
```diff
@@ -1,18 +1,11 @@
 import { instrument } from '@storybook/instrumenter';
-import * as spy from '@vitest/spy';
+import { type LoaderFunction } from '@storybook/csf';
 import chai from 'chai';
-import { FORCE_REMOUNT, STORY_RENDER_PHASE_CHANGED } from '@storybook/core-events';
-import { addons } from '@storybook/preview-api';
+import { global } from '@storybook/global';
```
Comment (yannbf) [inline code/lib/test/src/index.ts:4]: you should add this as a dependency in the package

File: code/lib/preview-api/src/modules/store/csf/prepareStory.ts [hunk 4]
```diff
@@ -48,15 +50,23 @@ export function prepareStory<TRenderer extends Renderer>(
...
+    context: StoryContextForLoaders<TRenderer>
+  ): Promise<StoryContextForLoaders<TRenderer> & { loaded: StoryContext<TRenderer>['loaded'] }> => {
+    let updatedContext = { ...context, loaded: {} };
+    for (const loaders of [
+      ...('testPackageLoaders' in global && Array.isArray(global.testPackageLoaders)
+        ? [global.testPackageLoaders]
+        : []),
+      normalizeArrays(projectAnnotations.loaders),
+      normalizeArrays(componentAnnotations.loaders),
+      normalizeArrays(storyAnnotations.loaders),
```
Comment (yannbf) [inline code/lib/preview-api/src/modules/store/csf/prepareStory.ts:63]: did you mean to make them `...normalizeArrays`?

File: code/lib/preview-api/src/modules/store/csf/prepareStory.ts [hunk 5]
```diff
@@ -48,15 +50,23 @@ export function prepareStory<TRenderer extends Renderer>(
...
-  const applyLoaders = async (context: StoryContextForLoaders<TRenderer>) => {
-    const loadResults = await Promise.all(loaders.map((loader) => loader(context)));
-    const loaded = Object.assign({}, ...loadResults);
-    return { ...context, loaded };
+  const applyLoaders = async (
+    context: StoryContextForLoaders<TRenderer>
+  ): Promise<StoryContextForLoaders<TRenderer> & { loaded: StoryContext<TRenderer>['loaded'] }> => {
+    let updatedContext = { ...context, loaded: {} };
+    for (const loaders of [
+      ...('testPackageLoaders' in global && Array.isArray(global.testPackageLoaders)
```
Comment (ndelangen) [inline code/lib/preview-api/src/modules/store/csf/prepareStory.ts:58]: Let's call this something like `__STORYBOOK_TEST_LOADERS__`.

File: code/lib/test/src/index.ts (hunk 2 above)
Comment (ndelangen) [reply code/lib/test/src/index.ts:39]: need FOR

**Review by ndelangen — CHANGES_REQUESTED** [review]

File: code/lib/preview-api/src/modules/store/csf/normalizeArrays.ts [hunk 6]
```diff
@@ -0,0 +1,4 @@
+export const normalizeArrays = <T>(array: T[] | T | undefined): T[] => {
+  if (Array.isArray(array)) return array;
+  return typeof array === 'function' ? [array] : [];
```
Comment (IanVS) [inline code/lib/preview-api/src/modules/store/csf/normalizeArrays.ts:3]: Is this only intended to be used for functions and arrays of functions?  Maybe the generic should be locked down just a bit, then?  Since it might be unexpected to get back `[]` from calling this with `normalizeArrays('foo')` (I'd expect to get back `['foo']`)

**Comment by kasperpeulen** [comment]
> Great job pulling this together @kasperpeulen. What's the release plan with `@storybook/csf`?

When I have CI green, and did some more QA, I think I will release csf to stable.

File: code/lib/preview-api/src/modules/store/csf/prepareStory.ts (hunk 4 above)
Comment (kasperpeulen) [reply code/lib/preview-api/src/modules/store/csf/prepareStory.ts:63]: Nope, because I want to sequentially await first the project loaders, then the component loaders, and then the story loaders!

File: code/lib/test/src/index.ts (hunk 3 above)
Comment (kasperpeulen) [reply code/lib/test/src/index.ts:4]: Check

File: code/lib/types/src/modules/story.ts (hunk 1 above)
Comment (kasperpeulen) [reply code/lib/types/src/modules/story.ts:52]: I change the type, the original type also allows non-arrays now.

File: code/lib/preview-api/src/modules/store/csf/normalizeArrays.ts (hunk 6 above)
Comment (kasperpeulen) [reply code/lib/preview-api/src/modules/store/csf/normalizeArrays.ts]: I think the function check is not neccesary actually, and we didn't do that before, so I just made the function more generic.

**Review by shilman — APPROVED** [review]
Amazing work! 💯
